// @piplabs/cdr-sdk/dist/story-api/decoders.js
/**
 * Raw JSON → typed shape decoders for /dkg/* responses.
 * Converts snake_case to camelCase, base64 to `Uint8Array`, numeric strings
 * to `bigint`.
 */
import { base64ToBytes } from "./bytes.js";
import { IncompleteDKGNetworkError } from "./errors.js";
// ---------------------------------------------------------------------------
// Decoders
// ---------------------------------------------------------------------------
export function decodeDKGNetwork(raw) {
    const missing = [];
    if (raw.start_block_height === undefined)
        missing.push("start_block_height");
    if (raw.start_block_hash === undefined)
        missing.push("start_block_hash");
    if (raw.active_val_set === undefined)
        missing.push("active_val_set");
    if (raw.global_public_key === undefined)
        missing.push("global_public_key");
    if (raw.public_coeffs === undefined)
        missing.push("public_coeffs");
    if (missing.length > 0) {
        throw new IncompleteDKGNetworkError(raw.round, raw.stage, missing);
    }
    return {
        round: raw.round,
        startBlockHeight: BigInt(raw.start_block_height),
        startBlockHash: base64ToBytes(raw.start_block_hash),
        activeValSet: raw.active_val_set.map((a) => a.toLowerCase()),
        total: raw.total,
        threshold: raw.threshold,
        stage: raw.stage,
        isResharing: raw.is_resharing,
        globalPublicKey: base64ToBytes(raw.global_public_key),
        publicCoeffs: raw.public_coeffs.map(base64ToBytes),
    };
}
export function decodeDKGRegistration(raw) {
    return {
        round: raw.round,
        validatorAddr: raw.validator_addr.toLowerCase(),
        index: raw.index,
        dkgPubKey: base64ToBytes(raw.dkg_pub_key),
        commPubKey: base64ToBytes(raw.comm_pub_key),
        pubKeyShare: raw.pub_key_share ? base64ToBytes(raw.pub_key_share) : undefined,
        enclaveReport: base64ToBytes(raw.enclave_report),
        status: raw.status,
        codeCommitment: base64ToBytes(raw.code_commitment),
        enclaveType: base64ToBytes(raw.enclave_type),
    };
}
export function decodeSubmission(raw) {
    return {
        validator: raw.validator.toLowerCase(),
        round: raw.round,
        pid: raw.pid,
        encryptedPartial: base64ToBytes(raw.encrypted_partial),
        ephemeralPubKey: base64ToBytes(raw.ephemeral_pub_key),
        pubShare: base64ToBytes(raw.pub_share),
        label: base64ToBytes(raw.label),
        ciphertext: base64ToBytes(raw.ciphertext),
    };
}
export function decodeSubmissionsByRound(raw) {
    return {
        round: raw.round,
        submissions: (raw.submissions ?? []).map(decodeSubmission),
        ciphertext: base64ToBytes(raw.ciphertext),
        threshold: raw.threshold,
        thresholdMet: raw.threshold_met,
    };
}
