// @piplabs/cdr-sdk/dist/story-api/errors.js
/**
 * Story-API REST client errors.
 * `StoryApiNotFoundError` for 404; `StoryApiError` for everything else.
 * `IncompleteDKGNetworkError` for non-Active/Ended rounds where the keeper
 * omits stage-conditional fields.
 */
export class StoryApiNotFoundError extends Error {
    constructor(message) {
        super(message);
        this.name = "StoryApiNotFoundError";
    }
}
export class StoryApiError extends Error {
    status;
    code;
    constructor(status, message, code) {
        super(message);
        this.name = "StoryApiError";
        this.status = status;
        this.code = code;
    }
}
/**
 * Thrown by `decodeDKGNetwork` when the keeper response omits one or more
 * stage-conditional fields. The keeper does not populate `global_public_key`,
 * `public_coeffs`, etc. for rounds in non-stable stages (Registration=1,
 * Dealing=2, Finalization=3, Failed=5). Callers that may query historical
 * or in-progress rounds should catch this and branch on stage; callers that
 * only ever read Active(4)/Ended(6) rounds will never see it in practice.
 */
export class IncompleteDKGNetworkError extends Error {
    round;
    stage;
    missingFields;
    constructor(round, stage, missingFields) {
        super(`DKG round ${round} (stage=${stage}) response is missing fields: ` +
            `${missingFields.join(", ")}. The keeper omits these for non-Active/Ended stages.`);
        this.name = "IncompleteDKGNetworkError";
        this.round = round;
        this.stage = stage;
        this.missingFields = missingFields;
    }
}
